{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import requests \n",
    "from bs4 import BeautifulSoup\n",
    "import time\n",
    "import pickle\n",
    "\n",
    "from dietrx import *\n",
    "from dietrx.models import *\n",
    "\n",
    "DATA = Path('data/version-3/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_colnames(df):\n",
    "    df.columns = [col.replace(' ', '-').lower() for col in df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_db(obj, dl):\n",
    "    for i, item in enumerate(dl):\n",
    "        db.session.add(obj(**item))\n",
    "\n",
    "        if (i+1) % 5000 == 0:\n",
    "            db.session.commit()\n",
    "    db.session.commit()\n",
    "    \n",
    "def add_to_db_df(obj, df):\n",
    "    for i, row in df.iterrows():\n",
    "        db.session.add(obj(**row.to_dict()))\n",
    "\n",
    "        if (i+1) % 5000 == 0:\n",
    "            db.session.commit()\n",
    "    db.session.commit()\n",
    "    \n",
    "    \n",
    "def remove_from_db(obj):\n",
    "    obj.query.delete()            \n",
    "    db.session.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove_from_db(Food_disease)\n",
    "# remove_from_db(Food_chemical)\n",
    "# remove_from_db(Food_gene)\n",
    "# remove_from_db(Disease_gene)\n",
    "# remove_from_db(Chemical_disease)\n",
    "# remove_from_db(Chemical_gene)\n",
    "# remove_from_db(Food)\n",
    "# remove_from_db(Disease)\n",
    "# remove_from_db(Gene)\n",
    "# remove_from_db(References)\n",
    "# remove_from_db(Chemical)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Vanilla Associations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### a. Food-Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food-id</th>\n",
       "      <th>disease-id</th>\n",
       "      <th>association</th>\n",
       "      <th>pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Plant ID:4045</td>\n",
       "      <td>MESH:D004487</td>\n",
       "      <td>negative</td>\n",
       "      <td>7398283</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         food-id    disease-id association     pmid\n",
       "0  Plant ID:4045  MESH:D004487    negative  7398283"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "food_dis = pd.read_csv(DATA/'food-disease.tsv', sep='\\t', encoding='utf-8')\n",
    "food_dis.drop_duplicates(['food-id', 'disease-id', 'association', 'pmid'], inplace=True)\n",
    "food_dis = food_dis.where((pd.notnull(food_dis)), None)\n",
    "food_dis['association'] = food_dis['association'].apply(str.lower)\n",
    "food_dis = food_dis.loc[:, ['food-id', 'disease-id', 'association', 'pmid']]\n",
    "food_dis['pmid'] = food_dis['pmid'].apply(str)\n",
    "food2dis = defaultdict(lambda: set(),\n",
    "                       food_dis.groupby('food-id').agg(lambda s: set(s))['disease-id'].to_dict())\n",
    "food_dis.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b. Food-Chemical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>content</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Plant ID:100170</td>\n",
       "      <td>5280378</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           food-id  pubchem-id                      content reference\n",
       "0  Plant ID:100170     5280378  Detected but not quantified  KNApSAcK"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "food_chem = pd.read_csv(DATA/'food-chemical.tsv', sep='\\t', encoding='utf-8')\n",
    "food2chem = defaultdict(lambda: set(),\n",
    "                        food_chem.groupby('food-id').agg(lambda s: set(s))['pubchem-id'].to_dict())\n",
    "\n",
    "food_chem.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### c. Gene-Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>gene-id</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000591739</td>\n",
       "      <td>1588</td>\n",
       "      <td>ORPHANET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C000596385</td>\n",
       "      <td>26504</td>\n",
       "      <td>UNIPROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:C000598645</td>\n",
       "      <td>9499</td>\n",
       "      <td>CTD_human;ORPHANET;UNIPROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:C000600608</td>\n",
       "      <td>29940</td>\n",
       "      <td>ORPHANET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C000600608</td>\n",
       "      <td>113189</td>\n",
       "      <td>CTD_human;ORPHANET;UNIPROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        disease-id  gene-id                      source\n",
       "0  MESH:C000591739     1588                    ORPHANET\n",
       "1  MESH:C000596385    26504                     UNIPROT\n",
       "2  MESH:C000598645     9499  CTD_human;ORPHANET;UNIPROT\n",
       "3  MESH:C000600608    29940                    ORPHANET\n",
       "4  MESH:C000600608   113189  CTD_human;ORPHANET;UNIPROT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_dis = pd.read_csv(DATA/'gene-disease.tsv', sep='\\t', encoding='utf-8', index_col='Unnamed: 0')\n",
    "dis2gene = defaultdict(lambda: set(),\n",
    "                        gene_dis.groupby('disease-id').agg(lambda s: set(s))['gene-id'].to_dict())\n",
    "gene2dis = defaultdict(lambda: set(),\n",
    "                        gene_dis.groupby('gene-id').agg(lambda s: set(s))['disease-id'].to_dict())\n",
    "\n",
    "gene_dis.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### d. Chemical-Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>disease-id</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>MESH:D007674</td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pubchem-id    disease-id         type\n",
       "0          19  MESH:D007674  therapeutic"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chem_dis = pd.read_csv(DATA/'chemical-disease.tsv', sep='\\t', encoding='utf-8', index_col='Unnamed: 0')\n",
    "chem_dis = chem_dis.groupby(['pubchem-id', 'disease-id']).agg(lambda s: sorted(s)[-1]).reset_index()\n",
    "dis2chem = defaultdict(lambda: set(),\n",
    "                       chem_dis.groupby('disease-id').agg(lambda s: set(s))['pubchem-id'].to_dict())\n",
    "chem2dis = defaultdict(lambda: set(),\n",
    "                       chem_dis.groupby('pubchem-id').agg(lambda s: set(s))['disease-id'].to_dict())\n",
    "\n",
    "chem_dis.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### e. Chemical-Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>gene-id</th>\n",
       "      <th>interaction-actions</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>351</td>\n",
       "      <td>affects^binding|decreases^reaction</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>4313</td>\n",
       "      <td>decreases^activity|decreases^reaction|decrease...</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51</td>\n",
       "      <td>3091</td>\n",
       "      <td>affects^binding|affects^cotreatment|increases^...</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>51</td>\n",
       "      <td>7428</td>\n",
       "      <td>affects^binding|affects^cotreatment|increases^...</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>51</td>\n",
       "      <td>8290</td>\n",
       "      <td>affects^cotreatment|decreases^methylation|incr...</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pubchem-id  gene-id                                interaction-actions  \\\n",
       "0          19      351                 affects^binding|decreases^reaction   \n",
       "1          19     4313  decreases^activity|decreases^reaction|decrease...   \n",
       "2          51     3091  affects^binding|affects^cotreatment|increases^...   \n",
       "3          51     7428  affects^binding|affects^cotreatment|increases^...   \n",
       "4          51     8290  affects^cotreatment|decreases^methylation|incr...   \n",
       "\n",
       "  source  \n",
       "0    CTD  \n",
       "1    CTD  \n",
       "2    CTD  \n",
       "3    CTD  \n",
       "4    CTD  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chem_gene = pd.read_csv(DATA/'chemical-gene.tsv', sep='\\t', encoding='utf-8')\n",
    "chem2gene = defaultdict(lambda: set(),\n",
    "                        chem_gene.groupby('pubchem-id').agg(lambda s: set(s))['gene-id'].to_dict())\n",
    "gene2chem = defaultdict(lambda: set(),\n",
    "                        chem_gene.groupby('gene-id').agg(lambda s: set(s))['pubchem-id'].to_dict())\n",
    "chem_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'affects', 'decreases', 'increases'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set([act.split('^')[0] for actionlst in chem_gene['interaction-actions'] for act in actionlst.split('|')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADP-ribosylation',\n",
       " 'N-linked glycosylation',\n",
       " 'O-linked glycosylation',\n",
       " 'abundance',\n",
       " 'acetylation',\n",
       " 'activity',\n",
       " 'alkylation',\n",
       " 'amination',\n",
       " 'binding',\n",
       " 'carboxylation',\n",
       " 'chemical synthesis',\n",
       " 'cleavage',\n",
       " 'cotreatment',\n",
       " 'degradation',\n",
       " 'export',\n",
       " 'expression',\n",
       " 'farnesylation',\n",
       " 'folding',\n",
       " 'geranoylation',\n",
       " 'glucuronidation',\n",
       " 'glutathionylation',\n",
       " 'glycation',\n",
       " 'glycosylation',\n",
       " 'hydrolysis',\n",
       " 'hydroxylation',\n",
       " 'import',\n",
       " 'lipidation',\n",
       " 'localization',\n",
       " 'metabolic processing',\n",
       " 'methylation',\n",
       " 'mutagenesis',\n",
       " 'nitrosation',\n",
       " 'oxidation',\n",
       " 'phosphorylation',\n",
       " 'prenylation',\n",
       " 'reaction',\n",
       " 'reduction',\n",
       " 'response to substance',\n",
       " 'secretion',\n",
       " 'splicing',\n",
       " 'stability',\n",
       " 'sulfation',\n",
       " 'sumoylation',\n",
       " 'transport',\n",
       " 'ubiquitination',\n",
       " 'uptake'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set([act.split('^')[1] for actionlst in chem_gene['interaction-actions'] for act in actionlst.split('|')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lexicons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A. Food "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>common_names</th>\n",
       "      <th>food_id</th>\n",
       "      <th>food_category</th>\n",
       "      <th>display_name</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>tax_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alexanders; horse parsley</td>\n",
       "      <td>Plant ID:40962</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Alexanders</td>\n",
       "      <td>Smyrnium olusatrum</td>\n",
       "      <td>40962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>Plant ID:942083</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Scrophularia umbrosa</td>\n",
       "      <td>Scrophularia umbrosa</td>\n",
       "      <td>942083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dotted blazing star</td>\n",
       "      <td>Plant ID:344074</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Dotted blazing star</td>\n",
       "      <td>Liatris punctata</td>\n",
       "      <td>344074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plymouth pear</td>\n",
       "      <td>Plant ID:761867</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Plymouth pear</td>\n",
       "      <td>Pyrus cordata</td>\n",
       "      <td>761867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>Plant ID:49166</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Rhododendron kaempferi</td>\n",
       "      <td>Rhododendron kaempferi</td>\n",
       "      <td>49166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                common_names          food_id  food_category  \\\n",
       "0  alexanders; horse parsley   Plant ID:40962  Miscellaneous   \n",
       "1                       None  Plant ID:942083  Miscellaneous   \n",
       "2        dotted blazing star  Plant ID:344074  Miscellaneous   \n",
       "3              plymouth pear  Plant ID:761867  Miscellaneous   \n",
       "4                       None   Plant ID:49166  Miscellaneous   \n",
       "\n",
       "             display_name         scientific_name  tax_id  \n",
       "0              Alexanders      Smyrnium olusatrum   40962  \n",
       "1    Scrophularia umbrosa    Scrophularia umbrosa  942083  \n",
       "2     Dotted blazing star        Liatris punctata  344074  \n",
       "3           Plymouth pear           Pyrus cordata  761867  \n",
       "4  Rhododendron kaempferi  Rhododendron kaempferi   49166  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read file\n",
    "foodlex = pd.read_csv(DATA/'food-lexicon-automated.tsv', sep='\\t', encoding='utf-8')\n",
    "clean_colnames(foodlex)\n",
    "\n",
    "del foodlex['foodb-group'], foodlex['foodb-subgroup']\n",
    "\n",
    "# Replace NaN with None\n",
    "foodlex = foodlex.where((pd.notnull(foodlex)), None)\n",
    "\n",
    "# Rename columns to match db schema.\n",
    "foodlex.rename(columns={\n",
    "    'common-names': 'common_names',\n",
    "    'food-name': 'display_name',\n",
    "    'food-id': 'food_id',\n",
    "    'food-category': 'food_category',\n",
    "    'scientific-name': 'scientific_name',\n",
    "    'tax-id':'tax_id',\n",
    "}, inplace=True)\n",
    "\n",
    "foodlex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redundancy check\n",
    "assert len((set(food_dis['food-id']) | set(food_chem['food-id'])) - set(foodlex['food_id'])) == 0,\\\n",
    "    'Food lexicon is not exhaustive!'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_list= list(foodlex.T.to_dict().values())\n",
    "add_to_db(Food, data_list)\n",
    "del data_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B. Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "b'Skipping line 5374: expected 9 fields, saw 10\\n'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>disease_category</th>\n",
       "      <th>disease_synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C538288</td>\n",
       "      <td>10p Deletion Syndrome (Partial)</td>\n",
       "      <td>Congenital abnormality|Genetic disease (inborn...</td>\n",
       "      <td>Chromosome 10, 10p- Partial|Chromosome 10, mon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C535484</td>\n",
       "      <td>13q deletion syndrome</td>\n",
       "      <td>Congenital abnormality|Genetic disease (inborn...</td>\n",
       "      <td>Chromosome 13q deletion|Chromosome 13q deletio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:C579849</td>\n",
       "      <td>15q24 Microdeletion</td>\n",
       "      <td>Congenital abnormality|Genetic disease (inborn...</td>\n",
       "      <td>15q24 Deletion|15q24 Microdeletion Syndrome|In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:C579850</td>\n",
       "      <td>16p11.2 Deletion Syndrome</td>\n",
       "      <td>Congenital abnormality|Genetic disease (inborn...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C567076</td>\n",
       "      <td>17,20-Lyase Deficiency, Isolated</td>\n",
       "      <td>Congenital abnormality|Endocrine system diseas...</td>\n",
       "      <td>17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease_id                      disease_name  \\\n",
       "0  MESH:C538288   10p Deletion Syndrome (Partial)   \n",
       "1  MESH:C535484             13q deletion syndrome   \n",
       "2  MESH:C579849               15q24 Microdeletion   \n",
       "3  MESH:C579850         16p11.2 Deletion Syndrome   \n",
       "4  MESH:C567076  17,20-Lyase Deficiency, Isolated   \n",
       "\n",
       "                                    disease_category  \\\n",
       "0  Congenital abnormality|Genetic disease (inborn...   \n",
       "1  Congenital abnormality|Genetic disease (inborn...   \n",
       "2  Congenital abnormality|Genetic disease (inborn...   \n",
       "3  Congenital abnormality|Genetic disease (inborn...   \n",
       "4  Congenital abnormality|Endocrine system diseas...   \n",
       "\n",
       "                                    disease_synonyms  \n",
       "0  Chromosome 10, 10p- Partial|Chromosome 10, mon...  \n",
       "1  Chromosome 13q deletion|Chromosome 13q deletio...  \n",
       "2  15q24 Deletion|15q24 Microdeletion Syndrome|In...  \n",
       "3                                               None  \n",
       "4  17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dislex = pd.read_csv(DATA/'CTD_diseases_new.tsv', sep='\\t', encoding='utf-8', error_bad_lines=False)\n",
    "\n",
    "# Keep only selected columns.\n",
    "dislex = dislex[['DiseaseID', 'DiseaseName', 'SlimMappings', 'Synonyms']]\n",
    "\n",
    "# Replace nan with None\n",
    "dislex = dislex.where((pd.notnull(dislex)), None)\n",
    "\n",
    "# Rename to match schema\n",
    "dislex.rename(columns={\n",
    "    'DiseaseID': 'disease_id',\n",
    "    'DiseaseName': 'disease_name',\n",
    "    'SlimMappings': 'disease_category',\n",
    "    'Synonyms': 'disease_synonyms'\n",
    "}, inplace=True)\n",
    "\n",
    "dislex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redundancy check\n",
    "assert len((set(food_dis['disease-id']) | set(chem_dis['disease-id']) | set(gene_dis['disease-id'])) \n",
    "           - set(dislex['disease_id'])) == 0,\\\n",
    "    'Disease lexicon is not exhaustive!'\n",
    "\n",
    "# Subset disease lexicon\n",
    "dislex = dislex.set_index('disease_id').\\\n",
    "    loc[list(set(food_dis['disease-id']) | set(chem_dis['disease-id']) | set(gene_dis['disease-id']))].\\\n",
    "    reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Disease, dislex)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C. Chemical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "chem = pd.read_csv(DATA/'chemical-lexicon.tsv', sep='\\t', encoding='utf-8').drop_duplicates(['pubchem_id'])\n",
    "\n",
    "# Replace nan with null values\n",
    "chem = chem.where((pd.notnull(chem)), None)\n",
    "\n",
    "# Rename columns\n",
    "chem.rename(columns={'canonical_smiles': 'smiles',},inplace=True)\n",
    "\n",
    "del chem['inchi'], chem['inchikey']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redundancy check\n",
    "assert len((set(food_chem['pubchem-id']) | set(chem_dis['pubchem-id']) | set(chem_gene['pubchem-id'])) \n",
    "           - set(chem['pubchem_id'])) == 0,\\\n",
    "    'Chemical lexicon is not exhaustive!'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0, len(chem), 50000):\n",
    "    data_list = list(chem.loc[i:50000-1+i, :].T.to_dict().values())\n",
    "    add_to_db(Chemical, data_list)\n",
    "    print(i)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D. Gene"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"rudraksht@iiitd.ac,in\"\n",
    "\n",
    "def retrieve_annotation(id_list):\n",
    "\n",
    "    \"\"\"Annotates Entrez Gene IDs using Bio.Entrez, in particular epost (to\n",
    "    submit the data to NCBI) and esummary to retrieve the information.\n",
    "    Returns a list of dictionaries with the annotations.\"\"\"\n",
    "\n",
    "    request = Entrez.epost(\"gene\",id=\",\".join(id_list))\n",
    "    result = Entrez.read(request)\n",
    "\n",
    "    webEnv = result[\"WebEnv\"]\n",
    "    queryKey = result[\"QueryKey\"]\n",
    "    data = Entrez.esummary(db=\"gene\", webenv=webEnv, query_key =\n",
    "            queryKey)\n",
    "    annotations = Entrez.read(data)\n",
    "\n",
    "    return annotations\n",
    "\n",
    "records = list()\n",
    "covered = list()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for i in range(0, len(genes), 100):\n",
    "    if i not in covered:\n",
    "        annotation = retrieve_annotation([str(g) for g in genes[i: i+100]])\n",
    "        records += annotation['DocumentSummarySet']['DocumentSummary']\n",
    "\n",
    "    if i % 1000 == 0:\n",
    "        print(\"Covered: \", len(records))\n",
    "        \n",
    "records = [dict(d) for d in records]\n",
    "for rec, geneid in zip(records, genes):\n",
    "    rec['gene_id'] = geneid\n",
    "\n",
    "pickle.dump(records, open('data/version-3/gene-lexicon-raw.p', 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>organism</th>\n",
       "      <th>other_symbols</th>\n",
       "      <th>synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A1B, ABG, GAB, HYST2477</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2MD, CPAMD5, FWP007, S863-7</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id               gene_name gene_symbol      organism  \\\n",
       "0        1  alpha-1-B glycoprotein        A1BG  Homo sapiens   \n",
       "1        2   alpha-2-macroglobulin         A2M  Homo sapiens   \n",
       "\n",
       "                  other_symbols  \\\n",
       "0       A1B, ABG, GAB, HYST2477   \n",
       "1  A2MD, CPAMD5, FWP007, S863-7   \n",
       "\n",
       "                                            synonyms  \n",
       "0  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...  \n",
       "1  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = pickle.load(open('data/version-3/gene-lexicon-raw.p', 'rb'))\n",
    "genelex = pd.DataFrame(records)\n",
    "\n",
    "# Subset\n",
    "genelex = genelex[['gene_id', 'Description', 'Name', 'Organism',\n",
    "                   'OtherAliases', 'OtherDesignations']]\n",
    "\n",
    "# Keep scientific name of organism only\n",
    "genelex['Organism'] = genelex['Organism'].map(lambda s: s['ScientificName'])\n",
    "\n",
    "# Rename columns\n",
    "genelex.rename(columns={\n",
    "    'Description': 'gene_name',\n",
    "    'Name': 'gene_symbol',\n",
    "    'Organism': 'organism',\n",
    "    'OtherAliases': 'other_symbols',\n",
    "    'OtherDesignations': 'synonyms'\n",
    "}, inplace=True)\n",
    "\n",
    "del records\n",
    "\n",
    "genelex.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redundancy check\n",
    "assert len((set(chem_gene['gene-id']) | set(gene_dis['gene-id'])) \n",
    "           - set(genelex['gene_id'])) == 0,\\\n",
    "    'Gene lexicon is not exhaustive!'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Gene, genelex)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### E. References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>authors</th>\n",
       "      <th>date</th>\n",
       "      <th>journal_name</th>\n",
       "      <th>journal_name_abbr</th>\n",
       "      <th>publication_type</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17344941</td>\n",
       "      <td>Elkiran T|Harputluoglu H|Yasar U|Babaoglu MO|D...</td>\n",
       "      <td>2007 Jan-Feb</td>\n",
       "      <td>Methods and findings in experimental and clini...</td>\n",
       "      <td>Methods Find Exp Clin Pharmacol</td>\n",
       "      <td>Clinical Trial|Journal Article</td>\n",
       "      <td>Differential alteration of drug-metabolizing e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25456153</td>\n",
       "      <td>Severins N|Mensink RP|Plat J</td>\n",
       "      <td>2015 Feb</td>\n",
       "      <td>Nutrition, metabolism, and cardiovascular dise...</td>\n",
       "      <td>Nutr Metab Cardiovasc Dis</td>\n",
       "      <td>Journal Article|Randomized Controlled Trial</td>\n",
       "      <td>Effects of lutein-enriched egg yolk in butterm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>23922960</td>\n",
       "      <td>Li C|Zuo C|Deng G|Kuang R|Yang Q|Hu C|Sheng O|...</td>\n",
       "      <td>2013</td>\n",
       "      <td>PloS one</td>\n",
       "      <td>PLoS One</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Contamination of bananas with beauvericin and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8766742</td>\n",
       "      <td>Güneser S|Atici A|Cengizler I|Alparslan N</td>\n",
       "      <td>1996 May-Jun</td>\n",
       "      <td>Allergologia et immunopathologia</td>\n",
       "      <td>Allergol Immunopathol (Madr)</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Inhalant allergens: as a cause of respiratory ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19918387</td>\n",
       "      <td>Brvar M|Bunc M</td>\n",
       "      <td>2009 Sep 9</td>\n",
       "      <td>Cases journal</td>\n",
       "      <td>Cases J</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>High-degree atrioventricular block in acute et...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid                                            authors          date  \\\n",
       "0  17344941  Elkiran T|Harputluoglu H|Yasar U|Babaoglu MO|D...  2007 Jan-Feb   \n",
       "1  25456153                       Severins N|Mensink RP|Plat J      2015 Feb   \n",
       "2  23922960  Li C|Zuo C|Deng G|Kuang R|Yang Q|Hu C|Sheng O|...          2013   \n",
       "3   8766742          Güneser S|Atici A|Cengizler I|Alparslan N  1996 May-Jun   \n",
       "4  19918387                                     Brvar M|Bunc M    2009 Sep 9   \n",
       "\n",
       "                                        journal_name  \\\n",
       "0  Methods and findings in experimental and clini...   \n",
       "1  Nutrition, metabolism, and cardiovascular dise...   \n",
       "2                                           PloS one   \n",
       "3                   Allergologia et immunopathologia   \n",
       "4                                      Cases journal   \n",
       "\n",
       "                 journal_name_abbr  \\\n",
       "0  Methods Find Exp Clin Pharmacol   \n",
       "1        Nutr Metab Cardiovasc Dis   \n",
       "2                         PLoS One   \n",
       "3     Allergol Immunopathol (Madr)   \n",
       "4                          Cases J   \n",
       "\n",
       "                              publication_type  \\\n",
       "0               Clinical Trial|Journal Article   \n",
       "1  Journal Article|Randomized Controlled Trial   \n",
       "2                              Journal Article   \n",
       "3                              Journal Article   \n",
       "4                              Journal Article   \n",
       "\n",
       "                                               title  \n",
       "0  Differential alteration of drug-metabolizing e...  \n",
       "1  Effects of lutein-enriched egg yolk in butterm...  \n",
       "2  Contamination of bananas with beauvericin and ...  \n",
       "3  Inhalant allergens: as a cause of respiratory ...  \n",
       "4  High-degree atrioventricular block in acute et...  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references = pd.read_csv(DATA/'publication-records.tsv', sep='\\t', encoding='utf-8')\n",
    "clean_colnames(references)\n",
    "\n",
    "references = references.where((pd.notnull(references)), None)\n",
    "\n",
    "references.rename(columns={\n",
    "    'authors': 'authors',\n",
    "    'journal-name': 'journal_name',\n",
    "    'journal-name-abbrv.': 'journal_name_abbr',\n",
    "    'publication-type':'publication_type',\n",
    "},inplace=True)\n",
    "\n",
    "pmids_not_mapped = list(str(s) for s in set(food_dis['pmid'].apply(int)) - set(references['pmid']))\n",
    "references = references.set_index('pmid').\\\n",
    "    reindex(index=references.pmid.tolist() + pmids_not_mapped).fillna('').\\\n",
    "    reset_index()\n",
    "\n",
    "references.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(References, references)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final Associations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Food-Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food-id</th>\n",
       "      <th>disease-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Plant ID:100170</td>\n",
       "      <td>MESH:D002583</td>\n",
       "      <td>5280378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Plant ID:100170</td>\n",
       "      <td>MESH:D058186</td>\n",
       "      <td>5280378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           food-id    disease-id pubchem-id\n",
       "0  Plant ID:100170  MESH:D002583    5280378\n",
       "1  Plant ID:100170  MESH:D058186    5280378"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Curated\n",
    "neg_fd = food_dis[food_dis.association == 'negative'].\\\n",
    "    groupby(['food-id', 'disease-id']).agg(lambda s: '|'.join(s))[['pmid']].reset_index(())\n",
    "neg_fd.reset_index(inplace=True)\n",
    "neg_fd.rename(columns={'pmid': 'negative-pmid'}, inplace=True)\n",
    "\n",
    "pos_fd = food_dis[food_dis.association == 'positive'].\\\n",
    "    groupby(['food-id', 'disease-id']).agg(lambda s: '|'.join(s))[['pmid']]\n",
    "pos_fd.reset_index(inplace=True)\n",
    "pos_fd.rename(columns={'pmid': 'positive-pmid'}, inplace=True)\n",
    "\n",
    "food_dis_curated = pos_fd.merge(neg_fd, on=['food-id', 'disease-id'], how='outer').fillna('')\n",
    "\n",
    "# Inferred\n",
    "fd_inferred = list()\n",
    "for food, chemicals in food2chem.items():\n",
    "    for chem in chemicals:\n",
    "        for dis in chem2dis[chem]:\n",
    "            fd_inferred.append([food, dis, str(chem)])\n",
    "            \n",
    "fd_inferred = pd.DataFrame(fd_inferred, \n",
    "                           columns=['food-id', 'disease-id', 'pubchem-id'])\n",
    "fd_inferred = fd_inferred.groupby(['food-id', 'disease-id']).agg(lambda s: '|'.join(s)).reset_index()\n",
    "del pos_fd, neg_fd\n",
    "\n",
    "fd_inferred.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food_id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>positive_pmid</th>\n",
       "      <th>negative_pmid</th>\n",
       "      <th>pubchem_id</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>MESH:D001523</td>\n",
       "      <td>7195588</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>MESH:D003866</td>\n",
       "      <td>6798614</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlcoholicBev ID:18</td>\n",
       "      <td>MESH:D005756</td>\n",
       "      <td>11848298</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlcoholicBev ID:18</td>\n",
       "      <td>MESH:D013272</td>\n",
       "      <td>17295722</td>\n",
       "      <td>9615879|17295722</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlcoholicBev ID:2</td>\n",
       "      <td>MESH:C537766</td>\n",
       "      <td>25604939</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              food_id    disease_id positive_pmid     negative_pmid  \\\n",
       "0  AlcoholicBev ID:17  MESH:D001523       7195588                     \n",
       "1  AlcoholicBev ID:17  MESH:D003866       6798614                     \n",
       "2  AlcoholicBev ID:18  MESH:D005756      11848298                     \n",
       "3  AlcoholicBev ID:18  MESH:D013272      17295722  9615879|17295722   \n",
       "4   AlcoholicBev ID:2  MESH:C537766      25604939                     \n",
       "\n",
       "  pubchem_id  weight  \n",
       "0                  2  \n",
       "1                  2  \n",
       "2                  2  \n",
       "3                  3  \n",
       "4                  2  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "food_dis_db = food_dis_curated.merge(fd_inferred, on=['food-id', 'disease-id'], how='outer').fillna('')\n",
    "food_dis_db['weight'] = food_dis_db.apply(lambda row: len(row['positive-pmid'].split('|')) +\n",
    "                                          len(row['negative-pmid'].split('|')), 1)\n",
    "del fd_inferred, food_dis_curated\n",
    "\n",
    "food_dis_db = food_dis_db.where((pd.notnull(food_dis_db)), None)\n",
    "\n",
    "food_dis_db.rename(columns={\n",
    "    'food-id': 'food_id',\n",
    "    'disease-id': 'disease_id',\n",
    "    'positive-pmid': 'positive_pmid',\n",
    "    'negative-pmid': 'negative_pmid',\n",
    "    'pubchem-id': 'pubchem_id',\n",
    "    'weight':'weight',\n",
    "}, inplace=True)\n",
    "\n",
    "food_dis_db.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Food_disease, food_dis_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Food-chemical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food_id</th>\n",
       "      <th>pubchem_id</th>\n",
       "      <th>content</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Plant ID:100170</td>\n",
       "      <td>5280378</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Plant ID:100170</td>\n",
       "      <td>44260113</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Plant ID:100506</td>\n",
       "      <td>1150</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Plant ID:100506</td>\n",
       "      <td>5202</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Plant ID:100506</td>\n",
       "      <td>5372945</td>\n",
       "      <td>Detected but not quantified</td>\n",
       "      <td>KNApSAcK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           food_id  pubchem_id                      content references\n",
       "0  Plant ID:100170     5280378  Detected but not quantified   KNApSAcK\n",
       "1  Plant ID:100170    44260113  Detected but not quantified   KNApSAcK\n",
       "2  Plant ID:100506        1150  Detected but not quantified   KNApSAcK\n",
       "3  Plant ID:100506        5202  Detected but not quantified   KNApSAcK\n",
       "4  Plant ID:100506     5372945  Detected but not quantified   KNApSAcK"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "food_chem_db = food_chem.where((pd.notnull(food_chem)), None)\n",
    "\n",
    "food_chem_db = food_chem_db.where((pd.notnull(food_chem_db)), None)\n",
    "\n",
    "food_chem_db.rename(columns={\n",
    "    'pubchem-id': 'pubchem_id',\n",
    "    'food-id': 'food_id',\n",
    "    'reference': 'references',\n",
    "    'type':'type_relation',\n",
    "    'inference-network':'inference_network'\n",
    "}, inplace=True)\n",
    "\n",
    "food_chem_db.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Food_chemical, food_chem_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Food-gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inferred through chemicals\n",
    "fg_inf_chem = list()\n",
    "for food, chemlst in food2chem.items():\n",
    "    for chem in chemlst:\n",
    "        for gene in chem2gene[chem]:\n",
    "            fg_inf_chem.append([food, str(chem), gene])\n",
    "            \n",
    "fg_inf_chem = pd.DataFrame(fg_inf_chem, columns=['food-id', 'pubchem-id', 'gene-id'])\n",
    "fg_inf_chem = fg_inf_chem.groupby(['food-id', 'gene-id']).agg(lambda s: '|'.join(s)).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inferred throuch diseases\n",
    "fg_inf_dis = list()\n",
    "for food, dislst in food2dis.items():\n",
    "    for dis in dislst:\n",
    "        for gene in dis2gene[dis]:\n",
    "            fg_inf_dis.append([food, dis, gene])\n",
    "            \n",
    "fg_inf_dis = pd.DataFrame(fg_inf_dis, columns=['food-id', 'disease-id', 'gene-id'])\n",
    "fg_inf_dis = fg_inf_dis.groupby(['food-id', 'gene-id']).agg(lambda s: '|'.join(s)).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>food_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>via_diseases</th>\n",
       "      <th>via_chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>2</td>\n",
       "      <td>MESH:D003866</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>15</td>\n",
       "      <td>MESH:D003866</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>52</td>\n",
       "      <td>MESH:D003866</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>81</td>\n",
       "      <td>MESH:D004487</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlcoholicBev ID:17</td>\n",
       "      <td>100</td>\n",
       "      <td>MESH:D004487</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              food_id  gene_id  via_diseases via_chemicals\n",
       "0  AlcoholicBev ID:17        2  MESH:D003866              \n",
       "1  AlcoholicBev ID:17       15  MESH:D003866              \n",
       "2  AlcoholicBev ID:17       52  MESH:D003866              \n",
       "3  AlcoholicBev ID:17       81  MESH:D004487              \n",
       "4  AlcoholicBev ID:17      100  MESH:D004487              "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "food_gene_db = fg_inf_dis.merge(fg_inf_chem, on=['food-id', 'gene-id'], how='outer').fillna('')\n",
    "food_gene_db.rename(columns={\n",
    "    'gene-id': 'gene_id',\n",
    "    'food-id': 'food_id',\n",
    "    'disease-id': 'via_diseases', \n",
    "    'pubchem-id': 'via_chemicals'}, inplace=True)\n",
    "\n",
    "del fg_inf_chem, fg_inf_dis\n",
    "\n",
    "# food_gene_db['weight'] = food_gene_db.apply(lambda row: len(row['via_diseases'].split('|')) +\n",
    "#                                             len(row['via_chemicals'].split('|')), 1)\n",
    "\n",
    "\n",
    "food_gene_db = food_gene_db.where((pd.notnull(food_gene_db)), None)\n",
    "\n",
    "food_gene_db.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# del food_gene_db['weight']\n",
    "add_to_db_df(Food_gene, food_gene_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "del food_gene_db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Disease Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>gene-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C531617</td>\n",
       "      <td>14</td>\n",
       "      <td>14985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C531617</td>\n",
       "      <td>35</td>\n",
       "      <td>14985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease-id  gene-id pubchem-id\n",
       "0  MESH:C531617       14      14985\n",
       "1  MESH:C531617       35      14985"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Curated\n",
    "dg_curated = gene_dis.copy(deep=True)\n",
    "\n",
    "# Inferred through chemicals\n",
    "dg_inferred = list()\n",
    "for disease, chemlst in dis2chem.items():\n",
    "    for chem in chemlst:\n",
    "        for gene in chem2gene[chem]:\n",
    "            dg_inferred.append([disease, str(chem), gene])\n",
    "            \n",
    "dg_inferred = pd.DataFrame(dg_inferred, columns=['disease-id', 'pubchem-id', 'gene-id'])\n",
    "dg_inferred = dg_inferred.groupby(['disease-id', 'gene-id']).agg(lambda s: '|'.join(s)).reset_index()\n",
    "dg_inferred.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>reference</th>\n",
       "      <th>via_chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000591739</td>\n",
       "      <td>1588</td>\n",
       "      <td>ORPHANET</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        disease_id  gene_id reference via_chemicals\n",
       "0  MESH:C000591739     1588  ORPHANET              "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge\n",
    "disease_gene_db = dg_curated.merge(dg_inferred, on=['disease-id', 'gene-id'], how='outer').fillna('')\n",
    "disease_gene_db.rename(columns={'pubchem-id': 'via-chemicals'}, inplace=True)\n",
    "del dg_inferred, dg_curated\n",
    "\n",
    "disease_gene_db = disease_gene_db.where((pd.notnull(disease_gene_db)), None)\n",
    "\n",
    "disease_gene_db.rename(columns={\n",
    "    'gene-id': 'gene_id',\n",
    "    'disease-id': 'disease_id',\n",
    "    'via-chemicals':'via_chemicals',\n",
    "    'source': 'reference'\n",
    "}, inplace=True)\n",
    "\n",
    "disease_gene_db.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Disease_gene, disease_gene_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Disease chemical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>gene-id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000591739</td>\n",
       "      <td>10680</td>\n",
       "      <td>1588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C000591739</td>\n",
       "      <td>14403</td>\n",
       "      <td>1588</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        disease-id pubchem-id gene-id\n",
       "0  MESH:C000591739      10680    1588\n",
       "1  MESH:C000591739      14403    1588"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Curated\n",
    "chem_dis_curated = chem_dis.copy(deep=True)\n",
    "chem_dis_curated['pubchem-id'] = chem_dis_curated['pubchem-id'].apply(str)\n",
    "\n",
    "# Inferred through genes\n",
    "dc_inferred = list()\n",
    "for disease, genelst in dis2gene.items():\n",
    "    for gene in genelst:\n",
    "        for chem in gene2chem[gene]:\n",
    "            dc_inferred.append([disease, str(chem), str(gene)])\n",
    "            \n",
    "dc_inferred = pd.DataFrame(dc_inferred, columns=['disease-id', 'pubchem-id', 'gene-id'])\n",
    "dc_inferred = dc_inferred.groupby(['disease-id', 'pubchem-id']).agg(lambda s: '|'.join(s)).reset_index()\n",
    "dc_inferred.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>via-genes</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000591739</td>\n",
       "      <td>10680</td>\n",
       "      <td>1588</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        disease-id pubchem-id via-genes type\n",
       "0  MESH:C000591739      10680      1588     "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge\n",
    "disease_chem_db = dc_inferred.merge(chem_dis_curated, on=['disease-id', 'pubchem-id'], how='outer').fillna('')\n",
    "disease_chem_db.rename(columns={'gene-id': 'via-genes'}, inplace=True)\n",
    "# del disease_chem_db['type']\n",
    "del dc_inferred, chem_dis_curated\n",
    "\n",
    "disease_chem_db.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# disease_chem_db['weight'] = disease_chem_db['via-genes'].apply(lambda s: len(s.split('|')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_chem_db = disease_chem_db.where((pd.notnull(disease_chem_db)), None)\n",
    "\n",
    "disease_chem_db.rename(columns={\n",
    "    'pubchem-id': 'pubchem_id',\n",
    "    'disease-id': 'disease_id',\n",
    "    'type':'type_relation',\n",
    "    'via-genes':'via_genes'\n",
    "},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "284612"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Chemical_disease.query.delete()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Chemical_disease, disease_chem_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Chemical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Curated\n",
    "chem_gene_curated = chem_gene.copy(deep=True)\n",
    "chem_gene_curated['gene-id'] = chem_gene_curated['gene-id'].apply(str)\n",
    "chem_gene_curated['pubchem-id'] = chem_gene_curated['pubchem-id'].apply(str)\n",
    "\n",
    "# Inferred through diseases\n",
    "gc_inferred = list()\n",
    "for gene, dislst in gene2dis.items():\n",
    "    for dis in dislst:\n",
    "        for chem in dis2chem[dis]:\n",
    "            gc_inferred.append([str(gene), str(chem), dis])\n",
    "            \n",
    "gc_inferred = pd.DataFrame(gc_inferred, columns=['gene-id', 'pubchem-id', 'disease-id'])\n",
    "gc_inferred = gc_inferred.groupby(['gene-id', 'pubchem-id']).agg(lambda s: '|'.join(s)).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge\n",
    "chem_gene_db = gc_inferred.merge(chem_gene_curated, on=['gene-id', 'pubchem-id'], how='outer').fillna('')\n",
    "chem_gene_db.rename(columns={'disease-id': 'via-diseases'}, inplace=True)\n",
    "del chem_gene_db['source']\n",
    "del gc_inferred, chem_gene_curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "chem_gene_db = chem_gene_db.where((pd.notnull(chem_gene_db)), None)\n",
    "\n",
    "chem_gene_db.rename(columns={\n",
    "    'pubchem-id': 'pubchem_id',\n",
    "    'gene-id': 'gene_id',\n",
    "    'via-diseases': 'via_diseases',\n",
    "    'interaction-actions': 'interaction_actions'\n",
    "}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_to_db_df(Chemical_gene, chem_gene_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chemical Images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem import MolFromSmiles, Draw\n",
    "from rdkit.Chem.AllChem import Compute2DCoords\n",
    "from shutil import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove '/dietrx/static/images/molecules/*': No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!rm /dietrx/static/images/molecules/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mol in Chemical.query.all():\n",
    "    try:\n",
    "        m = MolFromSmiles(mol.smiles)\n",
    "        tmp = Compute2DCoords(m)\n",
    "        Draw.MolToFile(m, 'dietrx/static/images/molecules/' + str(mol.pubchem_id) + '.png')\n",
    "    except:\n",
    "        print(mol.mol_id, mol.smiles)\n",
    "        copy('dietrx/static/images/no-image.png', \n",
    "             'dietrx/static/images/molecules/' + str(mol.pubchem_id) + '.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SDF Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "covered = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completed: 0\n",
      "Completed: 10\n",
      "Completed: 20\n",
      "Completed: 30\n",
      "Completed: 40\n",
      "Completed: 50\n",
      "Completed: 60\n",
      "Completed: 70\n",
      "Completed: 80\n",
      "Completed: 90\n",
      "Completed: 100\n",
      "Completed: 110\n",
      "Completed: 120\n",
      "Completed: 130\n",
      "Completed: 140\n",
      "Completed: 150\n",
      "Completed: 160\n",
      "Completed: 170\n",
      "Completed: 180\n",
      "Completed: 190\n",
      "Completed: 200\n",
      "Completed: 210\n",
      "Completed: 220\n",
      "Completed: 230\n",
      "Completed: 240\n",
      "Completed: 250\n",
      "Completed: 260\n",
      "Completed: 270\n",
      "Completed: 280\n",
      "Completed: 290\n",
      "Completed: 300\n",
      "Completed: 310\n",
      "Completed: 320\n",
      "Completed: 330\n",
      "Completed: 340\n",
      "Completed: 350\n",
      "Completed: 360\n",
      "Completed: 370\n",
      "Completed: 380\n",
      "Completed: 390\n",
      "Completed: 400\n",
      "Completed: 410\n",
      "Completed: 420\n",
      "Completed: 430\n",
      "Completed: 440\n",
      "Completed: 450\n",
      "Completed: 460\n",
      "Completed: 470\n",
      "Completed: 480\n",
      "Completed: 490\n",
      "Completed: 500\n",
      "Completed: 510\n",
      "Completed: 520\n",
      "Completed: 530\n",
      "Completed: 540\n",
      "Completed: 550\n",
      "Completed: 560\n",
      "Completed: 570\n",
      "Completed: 580\n",
      "Completed: 590\n",
      "Completed: 600\n",
      "Completed: 610\n",
      "Completed: 620\n",
      "Completed: 630\n",
      "Completed: 640\n",
      "Completed: 650\n",
      "Completed: 660\n",
      "Completed: 670\n",
      "Completed: 680\n",
      "Completed: 690\n",
      "Completed: 700\n",
      "Completed: 710\n",
      "Completed: 720\n",
      "Completed: 730\n",
      "Completed: 740\n",
      "Completed: 750\n",
      "Completed: 760\n",
      "Completed: 770\n",
      "Completed: 780\n",
      "Completed: 790\n",
      "Completed: 800\n",
      "Completed: 810\n",
      "Completed: 820\n",
      "Completed: 830\n",
      "Completed: 840\n",
      "Completed: 850\n",
      "Completed: 860\n",
      "Completed: 870\n",
      "Completed: 880\n",
      "Completed: 890\n",
      "Completed: 900\n",
      "Completed: 910\n",
      "Completed: 920\n",
      "Completed: 930\n",
      "Completed: 940\n",
      "Completed: 950\n",
      "Completed: 960\n",
      "Completed: 970\n",
      "Completed: 980\n",
      "Completed: 990\n",
      "Completed: 1000\n",
      "Completed: 1010\n",
      "Completed: 1020\n",
      "Completed: 1030\n",
      "Completed: 1040\n",
      "Completed: 1050\n",
      "Completed: 1060\n",
      "Completed: 1070\n",
      "Completed: 1080\n",
      "Completed: 1090\n",
      "Completed: 1100\n",
      "Completed: 1110\n",
      "Completed: 1120\n",
      "Completed: 1130\n",
      "Completed: 1140\n",
      "Completed: 1150\n",
      "Completed: 1160\n",
      "Completed: 1170\n",
      "Completed: 1180\n",
      "Completed: 1190\n",
      "Completed: 1200\n",
      "Completed: 1210\n",
      "Completed: 1220\n",
      "Completed: 1230\n",
      "Completed: 1240\n",
      "Completed: 1250\n",
      "Completed: 1260\n",
      "Completed: 1270\n",
      "Completed: 1280\n",
      "Completed: 1290\n",
      "Completed: 1300\n",
      "Completed: 1310\n",
      "Completed: 1320\n",
      "Completed: 1330\n",
      "Completed: 1340\n",
      "Completed: 1350\n",
      "Completed: 1360\n",
      "Completed: 1370\n",
      "Completed: 1380\n",
      "Completed: 1390\n",
      "Completed: 1400\n",
      "Completed: 1410\n",
      "Completed: 1420\n",
      "Completed: 1430\n",
      "Completed: 1440\n",
      "Completed: 1450\n",
      "Completed: 1460\n",
      "Completed: 1470\n",
      "Completed: 1480\n",
      "Completed: 1490\n",
      "Completed: 1500\n",
      "Completed: 1510\n",
      "Completed: 1520\n",
      "Completed: 1530\n",
      "Completed: 1540\n",
      "Completed: 1550\n",
      "Completed: 1560\n",
      "Completed: 1570\n",
      "Completed: 1580\n",
      "Completed: 1590\n",
      "Completed: 1600\n",
      "Completed: 1610\n",
      "Completed: 1620\n",
      "Completed: 1630\n",
      "Completed: 1640\n",
      "Completed: 1650\n",
      "Completed: 1660\n",
      "Completed: 1670\n",
      "Completed: 1680\n",
      "Completed: 1690\n",
      "Completed: 1700\n",
      "Completed: 1710\n",
      "Completed: 1720\n",
      "Completed: 1730\n",
      "Completed: 1740\n",
      "Completed: 1750\n",
      "Completed: 1760\n",
      "Completed: 1770\n",
      "Completed: 1780\n",
      "Completed: 1790\n",
      "Completed: 1800\n",
      "Completed: 1810\n",
      "Completed: 1820\n",
      "Completed: 1830\n",
      "Completed: 1840\n",
      "Completed: 1850\n",
      "Completed: 1860\n",
      "Completed: 1870\n",
      "Completed: 1880\n",
      "Completed: 1890\n",
      "Completed: 1900\n",
      "Completed: 1910\n",
      "Completed: 1920\n",
      "Completed: 1930\n",
      "Completed: 1940\n",
      "Completed: 1950\n",
      "Completed: 1960\n",
      "Completed: 1970\n",
      "Completed: 1980\n",
      "Completed: 1990\n",
      "Completed: 2000\n",
      "Completed: 2010\n",
      "Completed: 2020\n",
      "Completed: 2030\n",
      "Completed: 2040\n",
      "Completed: 2050\n",
      "Completed: 2060\n",
      "Completed: 2070\n",
      "Completed: 2080\n",
      "Completed: 2090\n",
      "Completed: 2100\n",
      "Completed: 2110\n",
      "Completed: 2120\n",
      "Completed: 2130\n",
      "Completed: 2140\n",
      "Completed: 2150\n",
      "Completed: 2160\n",
      "Completed: 2170\n",
      "Completed: 2180\n",
      "Completed: 2190\n",
      "Completed: 2200\n",
      "Completed: 2210\n",
      "Completed: 2220\n",
      "Completed: 2230\n",
      "Completed: 2240\n",
      "Completed: 2250\n",
      "Completed: 2260\n",
      "Completed: 2270\n",
      "Completed: 2280\n",
      "Completed: 2290\n",
      "Completed: 2300\n",
      "Completed: 2310\n",
      "Completed: 2320\n",
      "Completed: 2330\n",
      "Completed: 2340\n",
      "Completed: 2350\n",
      "Completed: 2360\n",
      "Completed: 2370\n",
      "Completed: 2380\n",
      "Completed: 2390\n",
      "Completed: 2400\n",
      "Completed: 2410\n",
      "Completed: 2420\n",
      "Completed: 2430\n",
      "Completed: 2440\n",
      "Completed: 2450\n",
      "Completed: 2460\n",
      "Completed: 2470\n",
      "Completed: 2480\n",
      "Completed: 2490\n",
      "Completed: 2500\n",
      "Completed: 2510\n",
      "Completed: 2520\n",
      "Completed: 2530\n",
      "Completed: 2540\n",
      "Completed: 2550\n",
      "Completed: 2560\n",
      "Completed: 2570\n"
     ]
    },
    {
     "ename": "SSLError",
     "evalue": "HTTPSConnectionPool(host='cactus.nci.nih.gov', port=443): Max retries exceeded with url: /chemical/structure/C(=CC(=O)O)C(=O)O/file?format=sdf&get3d=True (Caused by SSLError(SSLError(\"bad handshake: SysCallError(-1, 'Unexpected EOF')\",),))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSysCallError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py\u001b[0m in \u001b[0;36mwrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname)\u001b[0m\n\u001b[1;32m    440\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 441\u001b[0;31m                 \u001b[0mcnx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdo_handshake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    442\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mOpenSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mWantReadError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/OpenSSL/SSL.py\u001b[0m in \u001b[0;36mdo_handshake\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1805\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_lib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL_do_handshake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ssl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1806\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_ssl_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ssl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/OpenSSL/SSL.py\u001b[0m in \u001b[0;36m_raise_ssl_error\u001b[0;34m(self, ssl, result)\u001b[0m\n\u001b[1;32m   1538\u001b[0m                         \u001b[0;32mraise\u001b[0m \u001b[0mSysCallError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrorcode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1539\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mSysCallError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Unexpected EOF\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1540\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mSysCallError\u001b[0m: (-1, 'Unexpected EOF')",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSSLError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    600\u001b[0m                                                   \u001b[0mbody\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 601\u001b[0;31m                                                   chunked=chunked)\n\u001b[0m\u001b[1;32m    602\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    345\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 346\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_conn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    347\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSocketTimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBaseSSLError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m    849\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'sock'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# AppEngine might not have  `.sock`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 850\u001b[0;31m             \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    851\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/connection.py\u001b[0m in \u001b[0;36mconnect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    325\u001b[0m             \u001b[0mserver_hostname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhostname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m             ssl_context=context)\n\u001b[0m\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/util/ssl_.py\u001b[0m in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir)\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mHAS_SNI\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Platform-specific: OpenSSL with enabled SNI\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 329\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrap_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mserver_hostname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mserver_hostname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    330\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py\u001b[0m in \u001b[0;36mwrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname)\u001b[0m\n\u001b[1;32m    447\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mOpenSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 448\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mssl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bad handshake: %r'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    449\u001b[0m             \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mSSLError\u001b[0m: (\"bad handshake: SysCallError(-1, 'Unexpected EOF')\",)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    444\u001b[0m                     \u001b[0mretries\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_retries\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 445\u001b[0;31m                     \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    446\u001b[0m                 )\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    638\u001b[0m             retries = retries.increment(method, url, error=e, _pool=self,\n\u001b[0;32m--> 639\u001b[0;31m                                         _stacktrace=sys.exc_info()[2])\n\u001b[0m\u001b[1;32m    640\u001b[0m             \u001b[0mretries\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/urllib3/util/retry.py\u001b[0m in \u001b[0;36mincrement\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    387\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mnew_retry\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_exhausted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 388\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mMaxRetryError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_pool\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merror\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mResponseError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcause\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='cactus.nci.nih.gov', port=443): Max retries exceeded with url: /chemical/structure/C(=CC(=O)O)C(=O)O/file?format=sdf&get3d=True (Caused by SSLError(SSLError(\"bad handshake: SysCallError(-1, 'Unexpected EOF')\",),))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSSLError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-42-be3a861fa872>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmol\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mChemical\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcovered\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrequests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'https://cactus.nci.nih.gov/chemical/structure/%s/file?format=sdf&get3d=True'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mmol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msmiles\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'dietrx/static/sdf_files/%i.sdf'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mmol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpubchem_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m             \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/api.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefault\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'allow_redirects'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'get'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/api.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 58\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    510\u001b[0m         }\n\u001b[1;32m    511\u001b[0m         \u001b[0msend_kwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 512\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    513\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    514\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    620\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    621\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 622\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    623\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dietrx/lib/python3.6/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    509\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreason\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_SSLError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    510\u001b[0m                 \u001b[0;31m# This branch is for urllib3 v1.22 and later.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 511\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mSSLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    512\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    513\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mSSLError\u001b[0m: HTTPSConnectionPool(host='cactus.nci.nih.gov', port=443): Max retries exceeded with url: /chemical/structure/C(=CC(=O)O)C(=O)O/file?format=sdf&get3d=True (Caused by SSLError(SSLError(\"bad handshake: SysCallError(-1, 'Unexpected EOF')\",),))"
     ]
    }
   ],
   "source": [
    "for i, mol in enumerate(Chemical.query.all()):\n",
    "    if i not in covered:\n",
    "        r = requests.get('https://cactus.nci.nih.gov/chemical/structure/%s/file?format=sdf&get3d=True' % mol.smiles)\n",
    "        with open('dietrx/static/sdf_files/%i.sdf' % mol.pubchem_id, 'w') as f:\n",
    "            f.write(r.text)\n",
    "\n",
    "            covered.add(i)\n",
    "        \n",
    "    if i % 10 == 0:\n",
    "        print(\"Completed: %i\" %i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4183"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(disease_chem_db['via-genes'] == '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>gene-id</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [disease-id, gene-id, source]\n",
       "Index: []"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_dis[gene_dis['disease-id'] == 'MESH:D009369']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease-id</th>\n",
       "      <th>pubchem-id</th>\n",
       "      <th>via-genes</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>280454</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>137</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280541</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>241</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280937</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>896</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280963</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>935</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280999</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>938</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281139</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>1054</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281230</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>1130</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281249</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>1140</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281393</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>2236</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281415</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>2353</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281792</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5154</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281839</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5694</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281878</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5757</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281888</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5780</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282047</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6013</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282091</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6037</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282118</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6042</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282164</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6047</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282219</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6124</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282269</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6202</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282293</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6228</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282421</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6253</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282481</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6549</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282589</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>7500</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282618</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>7847</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282727</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>10494</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282730</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>10545</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282799</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>14985</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282821</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>16315</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282823</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>16590</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283651</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>448537</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283694</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>493570</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283705</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>636970</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283721</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>637566</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283749</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>638024</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283760</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>643820</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283766</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>667639</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283781</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>969516</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283802</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>1548910</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283824</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>1548943</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283981</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5280343</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284013</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5280489</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284031</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5280784</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284083</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5280961</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284093</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5281168</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284136</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5281605</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284226</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5282155</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284268</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5320315</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284307</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5353546</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284318</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5354495</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284341</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5360545</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284361</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5379713</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284387</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>5462224</td>\n",
       "      <td></td>\n",
       "      <td>marker/mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284412</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6122729</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284430</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6436722</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284434</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6473881</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284437</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>6475145</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284459</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>9963391</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284501</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>10918539</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284519</th>\n",
       "      <td>MESH:D009369</td>\n",
       "      <td>54670067</td>\n",
       "      <td></td>\n",
       "      <td>therapeutic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          disease-id pubchem-id via-genes              type\n",
       "280454  MESH:D009369        137            marker/mechanism\n",
       "280541  MESH:D009369        241            marker/mechanism\n",
       "280937  MESH:D009369        896                 therapeutic\n",
       "280963  MESH:D009369        935            marker/mechanism\n",
       "280999  MESH:D009369        938                 therapeutic\n",
       "281139  MESH:D009369       1054            marker/mechanism\n",
       "281230  MESH:D009369       1130            marker/mechanism\n",
       "281249  MESH:D009369       1140            marker/mechanism\n",
       "281393  MESH:D009369       2236            marker/mechanism\n",
       "281415  MESH:D009369       2353                 therapeutic\n",
       "281792  MESH:D009369       5154            marker/mechanism\n",
       "281839  MESH:D009369       5694                 therapeutic\n",
       "281878  MESH:D009369       5757            marker/mechanism\n",
       "281888  MESH:D009369       5780                 therapeutic\n",
       "282047  MESH:D009369       6013            marker/mechanism\n",
       "282091  MESH:D009369       6037            marker/mechanism\n",
       "282118  MESH:D009369       6042            marker/mechanism\n",
       "282164  MESH:D009369       6047            marker/mechanism\n",
       "282219  MESH:D009369       6124            marker/mechanism\n",
       "282269  MESH:D009369       6202            marker/mechanism\n",
       "282293  MESH:D009369       6228            marker/mechanism\n",
       "282421  MESH:D009369       6253                 therapeutic\n",
       "282481  MESH:D009369       6549                 therapeutic\n",
       "282589  MESH:D009369       7500            marker/mechanism\n",
       "282618  MESH:D009369       7847            marker/mechanism\n",
       "282727  MESH:D009369      10494                 therapeutic\n",
       "282730  MESH:D009369      10545                 therapeutic\n",
       "282799  MESH:D009369      14985            marker/mechanism\n",
       "282821  MESH:D009369      16315                 therapeutic\n",
       "282823  MESH:D009369      16590                 therapeutic\n",
       "...              ...        ...       ...               ...\n",
       "283651  MESH:D009369     448537            marker/mechanism\n",
       "283694  MESH:D009369     493570                 therapeutic\n",
       "283705  MESH:D009369     636970                 therapeutic\n",
       "283721  MESH:D009369     637566                 therapeutic\n",
       "283749  MESH:D009369     638024                 therapeutic\n",
       "283760  MESH:D009369     643820                 therapeutic\n",
       "283766  MESH:D009369     667639                 therapeutic\n",
       "283781  MESH:D009369     969516                 therapeutic\n",
       "283802  MESH:D009369    1548910                 therapeutic\n",
       "283824  MESH:D009369    1548943                 therapeutic\n",
       "283981  MESH:D009369    5280343                 therapeutic\n",
       "284013  MESH:D009369    5280489                 therapeutic\n",
       "284031  MESH:D009369    5280784                 therapeutic\n",
       "284083  MESH:D009369    5280961                 therapeutic\n",
       "284093  MESH:D009369    5281168            marker/mechanism\n",
       "284136  MESH:D009369    5281605                 therapeutic\n",
       "284226  MESH:D009369    5282155                 therapeutic\n",
       "284268  MESH:D009369    5320315                 therapeutic\n",
       "284307  MESH:D009369    5353546                 therapeutic\n",
       "284318  MESH:D009369    5354495            marker/mechanism\n",
       "284341  MESH:D009369    5360545            marker/mechanism\n",
       "284361  MESH:D009369    5379713                 therapeutic\n",
       "284387  MESH:D009369    5462224            marker/mechanism\n",
       "284412  MESH:D009369    6122729                 therapeutic\n",
       "284430  MESH:D009369    6436722                 therapeutic\n",
       "284434  MESH:D009369    6473881                 therapeutic\n",
       "284437  MESH:D009369    6475145                 therapeutic\n",
       "284459  MESH:D009369    9963391                 therapeutic\n",
       "284501  MESH:D009369   10918539                 therapeutic\n",
       "284519  MESH:D009369   54670067                 therapeutic\n",
       "\n",
       "[79 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_chem_db[disease_chem_db['disease-id'] == 'MESH:D009369']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refrest Elastic Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "app.elasticsearch.indices.delete(index='food', ignore=[400, 404])\n",
    "app.elasticsearch.indices.delete(index='disease', ignore=[400, 404])\n",
    "app.elasticsearch.indices.delete(index='gene', ignore=[400, 404])\n",
    "\n",
    "\n",
    "Food.reindex(\"food_id\")\n",
    "Disease.reindex(\"disease_id\")\n",
    "Gene.reindex(\"gene_id\")\n",
    "\n",
    "app.elasticsearch.indices.delete(index='chemical', ignore=[400, 404])\n",
    "\n",
    "Chemical.reindex(\"pubchem_id\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dietrx",
   "language": "python",
   "name": "dietrx"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
